internal/handlers/common/compare.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	"golang.org/x/exp/constraints"
+
+	"github.com/FerretDB/FerretDB/internal/types"
+)
+
+// compareResult represents the result of a comparison.
+type compareResult int
+
+const (
+	equal compareResult = iota
+	less
+	greater
+	notEqual // but not less or greater; for example, two NaNs
+)
+
+// compareScalars compares two scalar values.
+func compareScalars(a, b any) compareResult {
+	if a == nil {
+		panic("a is nil")
+	}
+	if b == nil {
+		panic("b is nil")
+	}
+
+	switch a := a.(type) {
+	case float64:
+		switch b := b.(type) {
+		case float64:
```

Comment:
what about float32?

Reply:
float32 is not a BSON type

Reply:
See types package and https://bsonspec.org/spec.html